File: apps/api-gateway/src/infrastructure/health_checker.ts
/**
 * API Gateway Infrastructure — HTTP Health Checker
 *
 * Concrete implementation of HealthCheckPort.
 * Uses fetch with AbortSignal.timeout to enforce per-service timeout.
 */

import type { HealthCheckPort } from '../domain/repositories.js';
import type { ServiceConfig, ServiceHealthResult } from '../domain/models.js';

/** Extended config: virtual/alias services skip live health probes. */
export type ServiceConfigExtended = ServiceConfig & { noProbe?: boolean };

export class HTTPHealthChecker implements HealthCheckPort {
  async checkHealth(service: ServiceConfig): Promise<ServiceHealthResult> {
    const start = Date.now();
    const url = `${service.baseUrl}${service.healthPath}`;

    try {
      const resp = await fetch(url, {
        signal: AbortSignal.timeout(service.timeoutMs),
      });
      const latencyMs = Date.now() - start;

      if (resp.ok) {
        return { service: service.name, status: 'ok', latencyMs };
      }
      return {
        service: service.name,
        status: 'degraded',
        latencyMs,
        error: `HTTP ${resp.status}`,
      };
    } catch (err) {
      return {
        service: service.name,
        status: 'down',
        latencyMs: Date.now() - start,
        error: String(err),
      };
    }
  }
}

/** Static in-memory service registry. */
export class StaticServiceRegistry {
  private readonly services: ReturnType<typeof buildServiceConfigs>;

  constructor(serviceUrls: Record<string, string>) {
    this.services = buildServiceConfigs(serviceUrls);
  }

  /**
   * Returns services eligible for active health probing.
   * Virtual alias services (noProbe=true) are excluded — they share an
   * upstream with a real service and do not expose a separate /health endpoint.
   */
  getAllServices() {
    return Object.values(this.services).filter((s) => !s.noProbe);
  }

  getService(name: string) {
    return this.services[name];
  }
}

function buildServiceConfigs(urls: Record<string, string>) {
  const timeout = 2000;
  return {
    mcp:        { name: 'mcp',        baseUrl: urls['mcp']        ?? 'http://mcp-server:3000',          healthPath: '/health', timeoutMs: timeout },
    pdf:        { name: 'pdf',        baseUrl: urls['pdf']        ?? 'http://pdf-extractor:5001',       healthPath: '/health', timeoutMs: timeout },
    rag:        { name: 'rag',        baseUrl: urls['rag']        ?? 'http://rag-service:5002',         healthPath: '/health', timeoutMs: timeout },
    ta:         { name: 'ta',         baseUrl: urls['ta']         ?? 'http://technical-analysis:5003',  healthPath: '/health', timeoutMs: timeout },
    macro:      { name: 'macro',      baseUrl: urls['macro']      ?? 'http://macro-indicators:5004',    healthPath: '/health', timeoutMs: timeout },
    stock:      { name: 'stock',      baseUrl: urls['stock']      ?? 'http://stock-price:5000',         healthPath: '/health', timeoutMs: timeout },
    'kinh-dich': { name: 'kinh-dich', baseUrl: urls['kinh-dich'] ?? 'http://kinh-dich-service:5005',  healthPath: '/health', timeoutMs: timeout },
    alert:      { name: 'alert',      baseUrl: urls['alert']      ?? 'http://alert-engine:5006',        healthPath: '/health', timeoutMs: timeout },
    news:       { name: 'news',       baseUrl: urls['news']       ?? 'http://news-fetch:5008',           healthPath: '/health', timeoutMs: timeout },
    // Virtual alias: /api/* routes to MCP server with full path preserved.
    // noProbe=true: excluded from active health probes (no /health endpoint to ping separately).
    api:        { name: 'api',        baseUrl: urls['api']        ?? urls['mcp'] ?? 'http://mcp-server:3000', healthPath: '/health', timeoutMs: timeout, noProbe: true },
  } as Record<string, ServiceConfigExtended>;
}

File: apps/api-gateway-go/pkg/infrastructure/healthchecker.go
package infrastructure

import (
	"context"
	"fmt"
	"net/http"
	"time"

	"github.com/vn-market-intelligence/api-gateway-go/pkg/domain"
)

// HTTPHealthChecker is the concrete implementation of HealthCheckerPort.
// Uses net/http with context deadline for per-service timeout.
type HTTPHealthChecker struct {
	client *http.Client
}

// NewHTTPHealthChecker creates a new HTTPHealthChecker.
func NewHTTPHealthChecker() *HTTPHealthChecker {
	return &HTTPHealthChecker{
		client: &http.Client{},
	}
}

// CheckHealth checks the health of a single service.
func (h *HTTPHealthChecker) CheckHealth(ctx context.Context, svc *domain.ServiceConfig) (*domain.ServiceHealthResult, error) {
	start := time.Now()
	url := svc.BaseURL + svc.HealthPath

	timeoutCtx, cancel := context.WithTimeout(ctx, time.Duration(svc.TimeoutMs)*time.Millisecond)
	defer cancel()

	req, err := http.NewRequestWithContext(timeoutCtx, http.MethodGet, url, nil)
	if err != nil {
		return &domain.ServiceHealthResult{
			Service:   svc.Name,
			Status:    domain.StatusDown,
			LatencyMs: time.Since(start).Milliseconds(),
			Error:     err.Error(),
		}, nil
	}

	resp, err := h.client.Do(req)
	latencyMs := time.Since(start).Milliseconds()

	if err != nil {
		return &domain.ServiceHealthResult{
			Service:   svc.Name,
			Status:    domain.StatusDown,
			LatencyMs: latencyMs,
			Error:     err.Error(),
		}, nil
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		return &domain.ServiceHealthResult{
			Service:   svc.Name,
			Status:    domain.StatusOk,
			LatencyMs: latencyMs,
		}, nil
	}

	return &domain.ServiceHealthResult{
		Service:   svc.Name,
		Status:    domain.StatusDegraded,
		LatencyMs: latencyMs,
		Error:     fmt.Sprintf("HTTP %d", resp.StatusCode),
	}, nil
}
